src/Microsoft.DiaSymReader.PortablePdb/SymDocument.cs

```diff
@@ -136,9 +144,50 @@ public int GetSourceLength(out int length)
             out int count,
             [In, MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 4), Out]byte[] source)
         {
-            // SymReader doesn't support embedded source.
             count = 0;
-            return HResult.E_NOTIMPL;
+
+            // This function used to return E_NOTIMPL in all implementations. When filling in the
+            // implementation, it was deemed not very useful and impractical to actually decompress
+            // and decode text and interpret lines and columns here. A convention was adopted that 
+            // (0, 0, >= int.MaxValue, >= int.MaxValue) is a maximal range indicating that all
+            // source bytes should be returned. Anything else is rejected as invalid. This matches 
+            // the new native behavior.
+            if (startLine != 0 || 
+                startColumn != 0 ||
+                unchecked((uint)endLine) < int.MaxValue ||
+                unchecked((uint)endColumn) < int.MaxValue)
+            {
+                return HResult.E_INVALIDARG;
+            }
+
+            if (bufferLength < 0)
+            {
+                return HResult.E_INVALIDARG;
+            }
+
+            if (source == null && bufferLength > 0)
+            {
+                return HResult.E_INVALIDARG;
+            }
+
+            BlobReader reader;
+            int hr = GetEmbeddedSourceBlobReader(out reader);
+            if (hr != HResult.S_OK)
+            {
+                return hr;
+            }
+
+            count = Math.Min(bufferLength, reader.RemainingBytes);
+            if (count > 0)
+            {
+                // There is not currently a mechanism for reading from BlobReader to existing byte[]
+                // https://github.com/dotnet/corefx/issues/8004 tracks adding that API to corefx and
+                // this should be updated to use that when it's fixed. In the meantime, we are forced
+                // to make an extra copy here.
+                Array.Copy(reader.ReadBytes(count), source, count);
```

Comment:
Could we just add that one API now? Seems like making an extra copy here would be pretty significant hit.